File: scala/node/src/main/scala/com/giraffechain/consensus/Consensus.scala
package com.giraffechain.consensus

import cats.effect.implicits.*
import cats.effect.{Async, Resource}
import cats.implicits.*
import com.giraffechain.*
import com.giraffechain.codecs.{*, given}
import com.giraffechain.crypto.CryptoResources
import com.giraffechain.models.*
import com.giraffechain.utility.{Exp, Log1P}
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger

case class Consensus[F[_]](
    headerValidation: HeaderValidation[F],
    chainSelection: ChainSelection[F],
    stakerTracker: StakerTracker[F],
    etaCalculation: EtaCalculation[F],
    leaderElection: LeaderElection[F],
    localChain: LocalChain[F]
)

object Consensus:
  def make[F[_]: Async: CryptoResources](
      genesis: FullBlock,
      clock: Clock[F],
      dataStores: DataStores[F],
      eventIdGetterSetters: EventIdGetterSetters[F],
      blockIdTree: BlockIdTree[F],
      blockHeights: BlockHeights.BSS[F]
  ): Resource[F, Consensus[F]] =
    for {
      logger <- Slf4jLogger.fromName("Consensus").toResource
      given Logger[F] = logger
      protocolSettings <- Resource.pure[F, ProtocolSettings](ProtocolSettings.Default.merge(genesis.header.settings))
      etaCalculation <- EtaCalculation
        .make[F](dataStores.headers.getOrRaise, clock, genesis.header.stakerCertificate.eta.decodeBase58)
      epochBoundariesBSS <- EpochBoundaries.make[F](
        dataStores.epochBoundaries.pure[F],
        eventIdGetterSetters.epochBoundaries.get(),
        blockIdTree,
        eventIdGetterSetters.epochBoundaries.set,
        clock,
        dataStores.headers.getOrRaise
      )
      stakerDataBSS <- StakerData.make[F](
        StakerData.State(dataStores.activeStake, dataStores.inactiveStake, dataStores.stakers).pure[F],
        eventIdGetterSetters.stakerData.get(),
        blockIdTree,
        eventIdGetterSetters.stakerData.set,
        dataStores.bodies.getOrRaise,
        dataStores.transactions.getOrRaise,
        dataStores.transactionOutputs.getOrRaise
      )
      stakerTracker <- StakerTracker.make[F](clock, genesis.header.id, stakerDataBSS, epochBoundariesBSS)
      canonicalHeadId <- eventIdGetterSetters.canonicalHead.get().toResource
      canonicalHead <- dataStores.headers.getOrRaise(canonicalHeadId).toResource
      _ <- logger
        .info(
          show"Canonical head id=$canonicalHeadId height=${canonicalHead.height} slot=${canonicalHead.slot}"
        )
        .toResource
      localChain <- LocalChain.make[F](genesis.header.id, blockHeights, canonicalHeadId, dataStores.headers.getOrRaise)
      _ <- localChain.adoptions.evalTap(eventIdGetterSetters.canonicalHead.set).compile.drain.background
      _ <- adoptionLogger(localChain, dataStores)
      chainSelection <- ChainSelection.make[F](
        CryptoResources[F].blake2b512,
        CryptoResources[F].ed25519VRF,
        protocolSettings.chainSelectionKLookback,
        protocolSettings.chainSelectionSWindow
      )
      exp <- Exp.make()
      log1p <- Log1P.make() >>= Log1P.makeCached
      leaderElection <- LeaderElection.make(
        protocolSettings,
        CryptoResources[F].blake2b512,
        exp,
        log1p
      )
      headerValidation <- HeaderValidation.make[F](
        genesis.header.id,
        etaCalculation,
        stakerTracker,
        leaderElection,
        clock,
        dataStores.headers.getOrRaise
      )
    } yield Consensus(
      headerValidation,
      chainSelection,
      stakerTracker,
      etaCalculation,
      leaderElection,
      localChain
    )

  def adoptionLogger[F[_]: Async: Logger](localChain: LocalChain[F], dataStores: DataStores[F]) =
    localChain.adoptions
      .evalTap(id =>
        (dataStores.headers.getOrRaise(id), dataStores.bodies.getOrRaise(id)).tupled
          .flatMap((header, body) =>
            Logger[F].info(
              show"Adopted block id=${header.id} height=${header.height} slot=${header.slot} transactions=${body.transactionIds}"
            )
          )
      )
      .compile
      .drain
      .background

File: dart/protocol/lib/src/blockchain/consensus/consensus.dart
import 'package:giraffe_protocol/src/blockchain/consensus/chain_selection.dart';
import 'package:giraffe_sdk/sdk.dart';

import '../block_id_tree.dart';
import '../common/clock.dart';
import '../store.dart';
import '../codecs.dart';
import 'block_heights.dart';
import 'local_chain.dart';

export 'eta_calculation.dart';
export 'leader_election_validation.dart';
export 'numeric_utils.dart';
export 'staker_tracker.dart';
export 'utils.dart';
export 'models/vrf_argument.dart';

class Consensus {
  final LocalChain localChain;
  final ChainSelection chainSelection;

  Consensus({required this.localChain, required this.chainSelection});

  static Future<Consensus> make(
      FullBlock genesis,
      Clock clock,
      DataStores dataStores,
      BlockIdTree blockIdTree,
      ProtocolSettings protocolSettings,
      EventIdGetterSetters getterSetters) async {
    final blockHeightsBSS = BlockHeights.make(
        dataStores.blockHeightIndex,
        await getterSetters.blockHeightTree.get(),
        blockIdTree,
        getterSetters.blockHeightTree.set,
        dataStores.headers.getOrRaise);
    final localChain = LocalChainImpl(
        genesis: genesis.header.id,
        head: await getterSetters.canonicalHead.get(),
        fetchHeader: dataStores.headers.getOrRaise,
        fetchBody: dataStores.bodies.getOrRaise,
        blockHeightsBSS: blockHeightsBSS,
        onAdopted: getterSetters.canonicalHead.set);
    final chainSelection = ChainSelectionImpl(
        kLookback: protocolSettings.chainSelectionKLookback,
        sWindow: protocolSettings.chainSelectionSWindow);
    return Consensus(localChain: localChain, chainSelection: chainSelection);
  }

  Future<void> close() async {
    await localChain.close();
  }
}
